static/app/utils/useProjects.tsx

```diff
@@ -0,0 +1,416 @@
+  }
+
+  if (((!prevSearch && !search) || prevSearch === search) && cursor) {
+    query.cursor = cursor;
+  }
+
+  // "0" shouldn't be a valid value, so this check is fine
+  if (limit) {
+    query.per_page = limit;
+  }
+
+  if (allProjects) {
+    const {loading, projects} = ProjectsStore.getState();
+    // If the projects store is loaded then return all projects from the store
+    if (!loading) {
+      return {
+        results: projects,
+        hasMore: false,
+      };
+    }
+    // Otherwise mark the query to fetch all projects from the API
+    query.all_projects = 1;
+  }
+
+  let hasMore: null | boolean = false;
+  let nextCursor: null | string = null;
+  const [data, , resp] = await api.requestPromise(`/organizations/${orgId}/projects/`, {
+    includeAllArgs: true,
+    query,
+  });
+
+  const pageLinks = resp?.getResponseHeader('Link');
+  if (pageLinks) {
+    const paginationObject = parseLinkHeader(pageLinks);
+    hasMore =
+      paginationObject &&
+      (paginationObject.next.results || paginationObject.previous.results);
+    nextCursor = paginationObject.next.cursor;
+  }
+
+  // populate the projects store if all projects were fetched
+  if (allProjects) {
+    ProjectActions.loadProjects(data);
+  }
+
+  return {
+    results: data,
+    hasMore,
+    nextCursor,
+  };
+}
+
+/**
+ * Hook to efficently load projects.
+ */
+function useProjects({
+  orgId,
+  slugs,
+  limit,
+  allProjects,
+  passthroughPlaceholderProject = true,
+}: Options) {
+  const api = useApi();
+
+  const [state, setState] = useState<State>({
+    fetchedProjects: [],
+    projectsFromStore: [],
+    initiallyLoaded: false,
+    fetching: false,
+    isIncomplete: null,
+    hasMore: null,
+    prevSearch: null,
+    nextCursor: null,
+    fetchError: null,
+  });
+
+  // List of projects that need to be fetched via API
+  const fetchQueue = useRef(new Set<string>());
+
+  // Existing projects from the store
+  const [projects, setStoreProjects] = useState<Project[]>([]);
+
+  // listen for new projects from the store
+  useEffect(() => {
+    setStoreProjects(ProjectsStore.getAll());
+
+    const unsubscribe = ProjectsStore.listen(
+      () => setStoreProjects(ProjectsStore.getAll()),
+      undefined
+    );
+
+    return unsubscribe as () => void;
+  }, []);
+
+  useEffect(() => {
+    /**
+     * This will fetch projects via API (using project slug) provided by the fetchQueue
+     */
+    async function fetchSpecificProjects() {
+      if (!fetchQueue.current.size) {
+        return;
+      }
+
+      setState({...state, fetching: true});
+
+      let results: Project[] = [];
+      let fetchError = null;
+
+      const querySlugs = Array.from(fetchQueue.current);
+
+      try {
+        const resp = await fetchProjects(api, orgId, {slugs: querySlugs});
+        results = resp.results;
+      } catch (err) {
+        console.error(err); // eslint-disable-line no-console
```

Comment:
wouldn't it be better if we use `Sentry.captureException`?

Reply:
Maybe. This is just from the original `<Projects>` implementation though so 🤷